src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs

```diff
@@ -1104,6 +1102,23 @@ private static ImmutableArray<AssemblyIdentity> GetReferencedAssemblies(CSharpCo
 
                     Debug.Assert(sourceReferencedAssemblies.Length == sourceReferencedAssemblySymbols.Length);
 
+                    // Pre-calculate size to allow this code to only require a single ImmutableArray allocation.
+                    var builderSize = 0;
+                    for (int i = 0; i < sourceReferencedAssemblies.Length; i++)
+                    {
+                        if (!sourceReferencedAssemblySymbols[i].IsLinked)
+                        {
+                            builderSize++;
+                        }
+                    }
+
+                    for (int i = 1; i < modules.Length; i++)
+                    {
+                        builderSize += modules[i].GetReferencedAssemblies().Length;
+                    }
```

Comment:
This appears to be equivalent:

```cs
var builderSize = modules.Sum(static symbol => symbol.GetReferencedAssemblies().Length)
    - sourceReferencedAssemblySymbols.Count(static symbol => !symbol.IsLinked);
```

However, to me the subtraction and inversion of `IsLinked` feels strange. I would prefer this:

```cs
var builderSize = sourceReferencedAssemblySymbols.Count(static symbol => symbol.IsLinked)
    + modules.Sum(static (symbol, index) => index > 0 ? symbol.GetReferencedAssemblies().Length : 0);
```

The latter would require a new `Sum` overload being defined next to the current one in ImmutableArrayExtensions.cs:

```cs
public static int Sum<T>(this ImmutableArray<T> items, Func<T, int, int> selector)
{
    var sum = 0;
    for (var i = 0; i < items.Length; i++)
        sum += selector(items[i], i);

    return sum;
}
```

Reply:
I think the following might be clearer to me, but I'm not too opinionated.

```var builderSize = modules.Sum(static (module, index) =>
{
    if (index == 0)
        return module.GetReferencedAssemblySymbols().Count(static identity => !identity.IsLinked);
    else
        return module.GetReferencedAssemblies().Length;
});```

Reply:
Sure, that works for me